src/main/scala/mesosphere/marathon/core/launchqueue/impl/TaskLauncherActor.scala

```diff
@@ -233,26 +233,34 @@ private class TaskLauncherActor(
   }
 
   private[this] def receiveInstanceUpdate: Receive = {
-    case update: InstanceDeleted =>
-      logger.info(s"receiveInstanceUpdate: ${update.id} was deleted (${update.condition})")
-      // A) If the app has constraints, we need to reconsider offers that
-      // we already rejected. E.g. when a host:unique constraint prevented
-      // us to launch tasks on a particular node before, we need to reconsider offers
-      // of that node after a task on that node has died.
-      //
-      // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
-      if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
-        maybeOfferReviver.foreach(_.reviveOffers())
+    case update: InstanceUpdated =>
+      if (update.condition.isTerminal & update.instance.isScheduled) {
+        logger.info(s"receiveInstanceUpdate: ${update.id} is terminal (${update.condition}) and scheduled.")
+        // A) If the app has constraints, we need to reconsider offers that
+        // we already rejected. E.g. when a host:unique constraint prevented
+        // us to launch tasks on a particular node before, we need to reconsider offers
+        // of that node after a task on that node has died.
+        //
+        // B) If a reservation timed out, already rejected offers might become eligible for creating new reservations.
+        if (runSpec.constraints.nonEmpty || (runSpec.isResident && shouldLaunchInstances)) {
+          maybeOfferReviver.foreach(_.reviveOffers())
+        }
       }
       syncInstance(update.instance.instanceId)
       OfferMatcherRegistration.manageOfferMatcherStatus()
       sender() ! Done
 
-    case change: InstanceChange =>
-      syncInstance(change.instance.instanceId)
+    case update: InstanceDeleted =>
+      // if an instance was deleted, it's not needed anymore and we only have to remove it from the internal state
+      logger.info(s"${update.instance.instanceId} was deleted. Will remove from internal state.")
+      removeInstanceFromInternalState(update.instance.instanceId)
```

Comment:
Why not call `syncInstance` like we've done before?

Reply:
it seems like overkill... if we got `InstanceDeleted` and we trust instance tracker implementation, it should mean that it is not there...